Rename node-related parameters

jacobperron opened the pull request:
The 'node-' prefix seems redundant since the attribute is part of a "node" tag.
Add a deprecation warning in case the old name is used.

I couldn't find a reason why 'node-name' was preferred in https://github.com/ros2/launch_ros/pull/73.
I'm suggesting in this PR that we change it 'name', which seems more intuitive for me, especially coming from ROS 1.

jacobperron commented:
I also just noticed that the [migration guide](https://index.ros.org/doc/ros2/Tutorials/Launch-files-migration-guide/#node) uses the attribute "name", which is currently inconsistent with the implementation.

hidmic commented on launch_ros/launch_ros/actions/node.py:
```diff
@@ -235,6 +237,11 @@ def parse(cls, entity: Entity, parser: Parser):
+            warnings.warn(
+                "The attribute 'node-name' is deprecated. Use the attribute 'name' instead.")
+            kwargs['node_name'] = parser.parse_substitution(node_name)
+        node_name = entity.get_attr('name', optional=True)
```
@jacobperron FYI `name` is also an attribute for plain processes. This change means it will no longer be possible to set both independently (it currently is possible in Python). Probably not the most relevant feature, though if node name is given a suffix to make it anon, that'll show in logs from now on.

jacobperron replied:
I'd overlooked that `name` was being used for ExecuteProcess. I've proposed a solution in 61503c9. We avoid forwarding the 'name' attribute to ExecuteProcess and expose a new attribute for node, 'exec-name', that keeps the feature of setting the process node.

hidmic commented:
> which is currently inconsistent with the implementation.

You're absolutely right, good catch.

hidmic commented on launch_ros/launch_ros/actions/node.py:
```diff
@@ -244,6 +244,13 @@ def parse(cls, entity: Entity, parser: Parser):
+                del kwargs['name']
+        exec_name = entity.get_attr('exec-name', optional=True)
+        if exec_name is not None:
+            kwargs['name'] = parser.parse_substitution(exec_name)
```
@jacobperron hmm, while this is a fine patch, it grows the gap between Python and markup-based launch files. This is something we've discussed with @ivanpauno several times, how detrimental to standardization is the fact that launch frontend parsing support allows one to depart from the original action "signature". IMHO, we should apply this exact same change to the `Node` action constructor itself and then parsing function should just use it. WDYT?

ivanpauno replied:
The problem with changing the `Node` signature is that it breaks API in a confusing way:
if somebody was using `name` to specify the executable name, they will have a surprise when that's now used as node name.
I think will should keep `XML` attributes as close as possible to `Python` constructor arguments.

hidmic replied:
That is also true. But doing so in XML/YAML only but not in Python makes it worst IMO.

jacobperron replied:
I agree that this change should happen in the Node constructor, keeping the signature the same as the front-end. I'll do this if there's a consensus to go ahead with this change.

> The problem with changing the Node signature is that it breaks API in a confusing way:
if somebody was using name to specify the executable name, they will have a surprise when that's now used as node name.

Yeah, this is a problem. IMO, it seems like a good thing to change long term. In the short-term, we can document the behavior change in the release notes and on whatever documentation exists for launch_ros (note, besides the API docstrings, I only found mention of the "name" attribute in the [migration guide](https://index.ros.org/doc/ros2/Tutorials/Launch-files-migration-guide/#node) and it is already assuming the behavior I expect).

I also understand if there's consensus to leave the implementation as-is in order to avoid breaking the existing behavior.

ivanpauno replied:
I think that the API break is reasonable.
To avoid breaking user launch files, `node_name` can be deprecated with a warning.
The users that were using `name` will have to check the release notes to understand the problem (though I expect not much users using it).
We should also merge the argument name change together with a PR updating the release notes. If not, we will forget later.

If @wjwwood and @hidmic are also ok with breaking API, I think it's fine.

wjwwood replied:
👍 to breaking API to make it make sense, and 👍 for updating the documentation and release notes simultaneously.

jacobperron replied:
Thanks for the input, all. I'm on it :sunglasses:

jacobperron commented:
@hidmic @ivanpauno I've reverted the 'name' overriding in the parser (45d4992) and added new Node parameters for `name` and `namespace` (827042b), deprecating the old parameters. I've also added a new Node param `exec_name` for setting the process name (e7e1325)

ivanpauno commented on launch_ros/launch_ros/actions/node.py:
```diff
@@ -61,8 +63,11 @@ def __init__(
+        name: Optional[SomeSubstitutionsType] = None,
+        namespace: Optional[SomeSubstitutionsType] = '',
         node_name: Optional[SomeSubstitutionsType] = None,
         node_namespace: SomeSubstitutionsType = '',
```
Maybe the note in the docstring it's just fine.

jacobperron replied:
I've added a note to the param docstrings, hopefully readers will notice one of the two notes, and/or the deprecation warning printed to the terminal. (e069970)

ivanpauno commented on launch_ros/launch_ros/actions/node.py:
```diff
@@ -115,9 +120,16 @@ def __init__(
+        :param: exec_name the label used to represent the process.
+            Defaults to the basename of node executable.
         :param: node_name the name of the node
         :param: node_namespace the ros namespace for this Node
```
I would add a comment here indicating deprecation. So, it's more obvious for someone that jumped to the documentation of these arguments.

jacobperron replied:
Done e069970

ivanpauno commented on launch_ros/launch_ros/actions/node.py:
```diff
@@ -135,8 +147,18 @@ def __init__(
         cmd += ['--ros-args']  # Prepend ros specific arguments with --ros-args flag
         if node_name is not None:
+            warnings.warn("The parameter 'node_name' is deprecated, use 'name' instead")
+            if name is None:
```
Passing both `name` and `node_name` should be detected as an error IMO. I would add an else case.

jacobperron replied:
Raising an exception if this happens: fa4ab9f

ivanpauno commented on launch_ros/launch_ros/actions/node.py:
```diff
@@ -135,8 +147,18 @@ def __init__(
+        if node_namespace:
+            warnings.warn("The parameter 'node_namespace' is deprecated, use 'namespace' instead")
+            if not namespace:
+                namespace = node_namespace
```
same applies here

jacobperron replied:
See fa4ab9f

ivanpauno commented on launch_ros/launch_ros/actions/node.py:
```diff
@@ -237,13 +261,19 @@ def parse(cls, entity: Entity, parser: Parser):
             kwargs['node_name'] = parser.parse_substitution(node_name)
+        node_name = entity.get_attr('name', optional=True)
+        if node_name is not None:
+            kwargs['name'] = parser.parse_substitution(node_name)
```
If both `name` and `node_name` are specified, it should be detected as an error.

jacobperron replied:
This is detected by fa4ab9f

jacobperron commented:
> It would be good to check if something has to be updated in demos/examples/tutorials.

Good idea, I'll open up PRs connecting to this for instances I find.

jacobperron commented:
`LifecylceNode` has a required keyword argument named `node_name`. I've just patched how it is passed to it's parent constructor (586e09c), but what do you think about also renaming it's argument to `name`? This would be a harder API break, since we would would have to introduce a new argument `name` that is required.

jacobperron commented:
Adding to Foxy release notes: https://github.com/ros2/ros2_documentation/pull/522

ivanpauno commented:
> LifecylceNode has a required keyword argument named node_name. I've just patched how it is passed to it's parent constructor (586e09c), but what do you think about also renaming it's argument to name? This would be a harder API break, since we would would have to introduce a new argument name that is required.

I think that if we're going to do the change, it makes sense to rename all the `node_*` of any action/description (`LifecycleNode`, `ComposableNodeContainer`, `ComposableNode`, ...).
To avoid that hard API break in that case, the following can be done:
```python3
  def __init__(self, *, name: Optional[Text] = None, node_name: Optional[Text] = None, **kwargs) -> None:
```

If `node_name` is passed, a warning is showed.
If both `name` and `node_name` are passed, an error is raised.
If only `name` is passed, we're happy.

In another release, `node_name` can be removed and `name` be required.

jacobperron commented:
> If node_name is passed, a warning is showed.
If both name and node_name are passed, an error is raised.
If only name is passed, we're happy.

Also, if both are `None`, an error raised. I don't think the `LifecycleNode` action will work otherwise.

I can do this.

jacobperron commented:
I've also deprecated the node_name parameter for LifecycleNode (09aa4c5) and updated the connected demos PR.

jacobperron commented:
And I've done the same for ComposableNodeContainer (4f19781).

jacobperron commented:
I've created a ticket (https://github.com/ros2/launch_ros/issues/123) so we don't forget to remove the deprecated API post-Foxy.

ivanpauno commented on launch_ros/launch_ros/actions/node.py:
```diff
@@ -61,8 +63,11 @@ def __init__(
         self, *,
         node_executable: SomeSubstitutionsType,
```
Maybe, change this one to executable too?

jacobperron replied:
Sure, I don't mind doing this, but it will trigger a lot more changes on dependent packages.

Also, I noticed that the front-end uses slight different names for this and for the package name: `exec` and `pkg`. I don't think anything needs to be done about it, but just thought I'd point it out since the discrepancies between Python and the front-end are what motivated the current changes I'm making.

ivanpauno replied:
> Sure, I don't mind doing this, but it will trigger a lot more changes on dependent packages.

> Also, I noticed that the front-end uses slight different names for this and for the package name: exec and pkg. I don't think anything needs to be done about it, but just thought I'd point it out since the discrepancies between Python and the front-end are what motivated the current changes I'm making.

Yeah, that's true. I think that have some arguments prefixed with `node_` and some others not, is strange. I would avoid the `node_` prefix everywhere, though it can be split in different PRs to avoid handling an annoing number of PRs at the same time.

About the difference between the frontend and python, I prefer the short versions for the frontend (`pkg`, `exec`), as the other ones are much more verbose. I also think that the current Python names are ok.

jacobperron replied:
I'll open a follow-up set of PRs deprecating `node_executable`.

jacobperron replied:
I finally got around to doing this. Here's the PR: https://github.com/ros2/launch_ros/pull/140

jacobperron commented:
Updated ComposableNode (5849c05) and the demos PR.

hidmic commented on launch_ros/launch_ros/actions/composable_node_container.py:
```diff
@@ -34,8 +35,10 @@ class ComposableNodeContainer(Node):
-        node_name: SomeSubstitutionsType,
-        node_namespace: SomeSubstitutionsType,
+        name: SomeSubstitutionsType = None,
+        namespace: SomeSubstitutionsType = None,
```
@jacobperron nit: if these have `None` as default, type annotation should read `Optional[SomeSubstitutionsType]`

jacobperron replied:
Fixed 7c00297

hidmic commented on launch_ros/launch_ros/actions/lifecycle_node.py:
```diff
@@ -36,7 +37,7 @@
     """Action that executes a ROS lifecycle node."""
 
-    def __init__(self, *, node_name: Text, **kwargs) -> None:
+    def __init__(self, *, name: Text = None, node_name: Optional[Text] = None, **kwargs) -> None:
```
@jacobperron same about type annotation.

jacobperron replied:
7c00297

hidmic commented on launch_ros/launch_ros/actions/composable_node_container.py:
```diff
@@ -45,12 +48,35 @@ def __init__(
-        :param: node_name the name of the node, mandatory for full container node name resolution
-        :param: node_namespace the ros namespace for this Node, mandatory for full container node
-           name resolution
+        .. warning:: Parameters `node_name` and `node_namespace` are deprecated.
```
@jacobperron BTW, it turns out there's a [`deprecated` directive](http://www.sphinx-doc.org/en/master/usage/restructuredtext/directives.html#directive-deprecated) we can use.

jacobperron replied:
Using deprecated directive 50395ef
I wasn't sure what version number to put, so I used "Foxy".

jacobperron commented:
* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=9356)](http://ci.ros2.org/job/ci_linux/9356/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=5040)](http://ci.ros2.org/job/ci_linux-aarch64/5040/)
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=7663)](http://ci.ros2.org/job/ci_osx/7663/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=9277)](http://ci.ros2.org/job/ci_windows/9277/)
* Windows-container [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows-container&build=144)](http://ci.ros2.org/job/ci_windows-container/144/)

hidmic commented on test_launch_ros/.eggs/README.txt:
```diff
@@ -0,0 +1,6 @@
+This directory contains eggs that were downloaded by setuptools to build, test, and run plug-ins.
```
@jacobperron hmm do we need to keep `.eggs` around?

jacobperron replied:
Oops, removed (a836234)

hidmic commented on test_launch_ros/test/test_launch_ros/actions/test_node.py:
```diff
@@ -191,6 +192,13 @@ def test_create_node_with_invalid_parameters(self):
+            node_name='my_node', node_namespace='my_ns'
+        )
+        self._assert_launch_no_errors([node_action])
+
```
@jacobperron should we also ensure that it does throw if both `name` and `node_name` are provided? Same for namespace.

jacobperron replied:
Added tests c19dc08

ivanpauno commented:
@jacobperron is this ready to be merged?

jacobperron commented:
I was hesitating due to the CI failures. I've retriggered CI.

nuclearsandwich commented:
@jacobperron I haven't completed the first CI test but in local research https://github.com/ros2/ci/pull/399 is a workaround for the installation paths which should allow the tests to run.
